### PR TITLE
Add support for EasyDNS

### PR DESCRIPTION
**Pull request by jdharmon**
https://easydns.com/
http://docs.sandbox.rest.easydns.net/

### CODE REVIEW

**Comment by ldez** [comment]
Hello, in order for a PR adding a DNS provider to be accepted, you have to:
- [ ] add a description to your PR
- [ ] be able to maintain this provider
- [ ] have a homogeneous design with the other providers
- [ ] add tests ("live" and units)
https://github.com/xenolf/lego/blob/c938de68f262c057542d44bcb67ef961cefcc60e/providers/dns/cloudflare/cloudflare_test.go#L125-L151

```shell
make test
```
- [ ] add the provider to the [readme.md](https://github.com/xenolf/lego#dns-providers)
- [ ] add a [provider descriptor](https://github.com/xenolf/lego/blob/master/providers/dns/cloudflare/cloudflare.toml)
- [ ] generate CLI help and documentation
```bash
make generate-dns
```
- [ ] be able to do: _(and put the output of this command to a comment in your PR)_
```bash
rm -rf ./lego

./lego -m your@email.com --dns YOUR_PROVIDER_NAME -d *.example.com -d example.com -s https://acme-staging-v02.api.letsencrypt.org/directory run
```
Note the wildcard domain is important.
- [ ] pass the linter ([golangci-lint](https://github.com/golangci/golangci-lint#install) must be installed):
```shell
make checks
```

**Comment by jdharmon** [comment]
I am willing to maintain this.

I will add CLI output this weekend. Validation is currently failing because the TXT record is being overwritten by the simultaneous requests.

File: providers/dns/easydns/easydns_test.go [hunk 1]
```diff
@@ -0,0 +1,454 @@
...
+		})
+	}
+}
+
+func TestSplitFqdn(t *testing.T) {
+	testCases := []struct {
+		desc           string
+		fqdn           string
+		expectedHost   string
+		expectedDoamin string
```
Comment (ldez) [inline providers/dns/easydns/easydns_test.go]: `expectedDoamin` -> `expectedDomain`

File: providers/dns/easydns/easydns.toml [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+[Configuration]
+  [Configuration.Credentials]
+    EASYDNS_TOKEN = "API Token"
+    EASYDNS_KEY = "API Key"
+  [Configuration.Additional]
+    EASYDNS_HOSTNAME = "The hostname of the API Server"
+    EASYDNS_POLLING_INTERVAL = "Time between DNS propagation check"
+    EASYDNS_PROPAGATION_TIMEOUT = "Maximum waiting time for DNS propagation"
+    EASYDNS_TTL = "The TTL of the TXT record used for the DNS challenge"
+    EASYDNS_HTTP_TIMEOUT = "API request timeout"
```
Comment (ldez) [inline providers/dns/easydns/easydns.toml]: Could you add:
```

[Links]
  API = "http://docs.sandbox.rest.easydns.net"

```

File: providers/dns/easydns/easydns.go [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+
+	if responseMsg != nil {
+		return json.NewDecoder(response.Body).Decode(responseMsg)
+	}
+
+	return nil
+}
+
+func splitFqdn(fqdn string) (host, domain string) {
+	parts := strings.Split(fqdn, ".")
```
Comment (ldez) [inline providers/dns/easydns/easydns.go]:
```go
func splitFqdn(fqdn string) (host, domain string) {
	parts := dns.SplitDomainName(fqdn)
	length := len(parts)

	host = strings.Join(parts[0:length-2], ".")
	domain = strings.Join(parts[length-2:length], ".")
	return
}
```

File: providers/dns/easydns/easydns.go [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+	"path"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/go-acme/lego/challenge/dns01"
+	"github.com/go-acme/lego/platform/config/env"
+)
+
+const defaultHostname = "rest.easydns.net"
```
Comment (ldez) [inline providers/dns/easydns/easydns.go]:
```go
const defaultEndpoint = "https://rest.easydns.net"
```

File: providers/dns/easydns/easydns.go [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+		return fmt.Errorf("easydns: error getting zone record: %v", err)
+	}
+
+	if recordID == "" {
+		recordID, err = d.addRecord(apiDomain, record)
+		if err != nil {
+			return fmt.Errorf("easydns: error adding zone record: %v", err)
+		}
+	} else {
+		err := d.updateRecord(recordID, record)
```
Comment (ldez) [inline providers/dns/easydns/easydns.go]: I think it will not work as expected

File: providers/dns/easydns/easydns.go [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+	URL                *url.URL
+	HTTPClient         *http.Client
+	PropagationTimeout time.Duration
+	PollingInterval    time.Duration
+}
+
+// NewDefaultConfig returns a default configuration for the DNSProvider
+func NewDefaultConfig() *Config {
+	return &Config{
+		Hostname:           env.GetOrDefaultString("EASYDNS_HOSTNAME", defaultHostname),
```
Comment (ldez) [inline providers/dns/easydns/easydns.go]: Prefer `EASYDNS_ENDPOINT` and `Endpoint`

**Comment by ldez** [comment]
If you have some problem for the multiple domain support (CLI commands) don't hesitate to ask me questions.

**Comment by jdharmon** [comment]
I fixed the issue in my code to allow multiple TXT records to be set. However, I cannot pass the CLI validation because of what looks like a bug with EasyDNS--only one TXT record is returned at a time. 

I can verify using their API that both records exist, but only the first record shows in the DNS query. Immediately after the first record is cleaned up, the second record appears in DNS. See output below, where cleanup happens at 11:35:02, and the TXT record shows up in DNS at 11:35:05.

I've tried setting the propagation timeout to as high as 5 minutes, but it doesn't make a difference. Have you see this with other providers?

```
go run ./cmd/lego -m jason@jdharmon.com --dns easydns -d *.jdharmon.com -d jdharmon.com -s https://acme-staging-v02.api.letsencrypt.org/directory run
2019/04/28 11:33:51 [INFO] [*.jdharmon.com, jdharmon.com] acme: Obtaining bundled SAN certificate
2019/04/28 11:33:52 [INFO] [*.jdharmon.com] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz/VTF0luDbKrrLJ1FG21bQQZF9RYv3mGiLv-wr3emM8vY
2019/04/28 11:33:52 [INFO] [jdharmon.com] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz/jtwABhVmU_uyWu1urQhc1mQlOnNtXq735lGFl4DWp6Q
2019/04/28 11:33:52 [INFO] [*.jdharmon.com] acme: use dns-01 solver
2019/04/28 11:33:52 [INFO] [jdharmon.com] acme: Could not find solver for: tls-alpn-01
2019/04/28 11:33:52 [INFO] [jdharmon.com] acme: Could not find solver for: http-01
2019/04/28 11:33:52 [INFO] [jdharmon.com] acme: use dns-01 solver
2019/04/28 11:33:52 [INFO] [*.jdharmon.com] acme: Preparing to solve DNS-01
2019/04/28 11:33:53 [INFO] [jdharmon.com] acme: Preparing to solve DNS-01
2019/04/28 11:33:55 [INFO] [*.jdharmon.com] acme: Trying to solve DNS-01
2019/04/28 11:33:55 [INFO] [*.jdharmon.com] acme: Checking DNS record propagation using [google-public-dns-a.google.com:53 google-public-dns-b.google.com:53]
2019/04/28 11:33:55 [INFO] Wait for propagation [timeout: 1m0s, interval: 15s]
2019/04/28 11:34:01 [INFO] [*.jdharmon.com] The server validated our request
2019/04/28 11:34:01 [INFO] [jdharmon.com] acme: Trying to solve DNS-01
2019/04/28 11:34:01 [INFO] [jdharmon.com] acme: Checking DNS record propagation using [google-public-dns-a.google.com:53 google-public-dns-b.google.com:53]
2019/04/28 11:34:01 [INFO] Wait for propagation [timeout: 1m0s, interval: 15s]
2019/04/28 11:34:01 [INFO] [jdharmon.com] acme: Waiting for DNS record propagation.
2019/04/28 11:34:16 [INFO] [jdharmon.com] acme: Waiting for DNS record propagation.
2019/04/28 11:34:31 [INFO] [jdharmon.com] acme: Waiting for DNS record propagation.
2019/04/28 11:34:46 [INFO] [jdharmon.com] acme: Waiting for DNS record propagation.
2019/04/28 11:35:01 [INFO] [*.jdharmon.com] acme: Cleaning DNS-01 challenge
2019/04/28 11:35:02 [INFO] [jdharmon.com] acme: Cleaning DNS-01 challenge
2019/04/28 11:35:03 Could not obtain certificates:
        acme: Error -> One or more domains had a problem:
[jdharmon.com] time limit exceeded: last error: NS dns1.easydns.com. did not return the expected TXT record [fqdn: _acme-challenge.jdharmon.com., value: UjazXwbjQj77OHLMzuIoqZIIrRpiGtNUebXwCkz3xAw]: hFA93zC8_-woIM2_RmPoNEqFiDGiY0x0KXjv_D8TPf4
exit status 1
```

**During validation:**
```
; <<>> DiG 9.14.0 <<>> @google-public-dns-a.google.com _acme-challenge.jdharmon.com TXT
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 44425
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
;_acme-challenge.jdharmon.com.  IN      TXT

;; ANSWER SECTION:
_acme-challenge.jdharmon.com. 9 IN      TXT     "hFA93zC8_-woIM2_RmPoNEqFiDGiY0x0KXjv_D8TPf4"

;; Query time: 33 msec
;; SERVER: 8.8.8.8#53(8.8.8.8)
;; WHEN: Sun Apr 28 11:34:24 Eastern Daylight Time 2019
;; MSG SIZE  rcvd: 113
```
**Immediately after cleanup:**
```
; <<>> DiG 9.14.0 <<>> @google-public-dns-a.google.com _acme-challenge.jdharmon.com TXT
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 1390
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
;_acme-challenge.jdharmon.com.  IN      TXT

;; ANSWER SECTION:
_acme-challenge.jdharmon.com. 9 IN      TXT     "UjazXwbjQj77OHLMzuIoqZIIrRpiGtNUebXwCkz3xAw"

;; Query time: 59 msec
;; SERVER: 8.8.8.8#53(8.8.8.8)
;; WHEN: Sun Apr 28 11:35:05 Eastern Daylight Time 2019
;; MSG SIZE  rcvd: 113
```

**Comment by ldez** [comment]
could you add:
```

// Sequential All DNS challenges for this provider will be resolved sequentially.
// Returns the interval between each iteration.
func (d *DNSProvider) Sequential() time.Duration {
	return d.config.SequenceInterval
}

```

and retry the command

**Comment by jdharmon** [comment]
Thanks! Sequential requests work perfectly.

```
> go run ./cmd/lego -m jason@jdharmon.com --dns easydns -d *.jdharmon.com -d jdharmon.com -s https://acme-staging-v02.api.letsencrypt.org/directory run
2019/04/28 13:04:49 No key found for account jason@jdharmon.com. Generating a P384 key.
2019/04/28 13:04:49 Saved key to C:\Users\Jason\go\src\github.com\go-acme\lego\.lego\accounts\acme-staging-v02.api.letsencrypt.org\jason@jdharmon.com\keys\jason@jdharmon.com.key
2019/04/28 13:04:49 Please review the TOS at https://letsencrypt.org/documents/LE-SA-v1.2-November-15-2017.pdf
Do you accept the TOS? Y/n
y
2019/04/28 13:04:52 [INFO] acme: Registering account for jason@jdharmon.com
!!!! HEADS UP !!!!

                Your account credentials have been saved in your Let's Encrypt
                configuration directory at "C:\Users\Jason\go\src\github.com\go-acme\lego\.lego\accounts".
                You should make a secure backup of this folder now. This
                configuration directory will also contain certificates and
                private keys obtained from Let's Encrypt so making regular
                backups of this folder is ideal.2019/04/28 13:04:52 [INFO] [*.jdharmon.com, jdharmon.com] acme: Obtaining bundled SAN certificate
2019/04/28 13:04:54 [INFO] [*.jdharmon.com] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz/66-sHweqgoou9IAUxHlNavGlc3G4AeHsFI3NOhLAQuc
2019/04/28 13:04:54 [INFO] [jdharmon.com] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz/ua7GZKV18tcqy6N2pO-kjaPyb4PNdC3trO3dpp8ZOrA
2019/04/28 13:04:54 [INFO] [*.jdharmon.com] acme: use dns-01 solver
2019/04/28 13:04:54 [INFO] [jdharmon.com] acme: Could not find solver for: tls-alpn-01
2019/04/28 13:04:54 [INFO] [jdharmon.com] acme: Could not find solver for: http-01
2019/04/28 13:04:54 [INFO] [jdharmon.com] acme: use dns-01 solver
2019/04/28 13:04:54 [INFO] [*.jdharmon.com] acme: Preparing to solve DNS-01
2019/04/28 13:04:56 [INFO] [*.jdharmon.com] acme: Trying to solve DNS-01
2019/04/28 13:04:56 [INFO] [*.jdharmon.com] acme: Checking DNS record propagation using [google-public-dns-a.google.com:53 google-public-dns-b.google.com:53]
2019/04/28 13:04:56 [INFO] Wait for propagation [timeout: 1m0s, interval: 2s]
2019/04/28 13:05:02 [INFO] [*.jdharmon.com] The server validated our request
2019/04/28 13:05:02 [INFO] [*.jdharmon.com] acme: Cleaning DNS-01 challenge
2019/04/28 13:05:03 [INFO] sequence: wait for 1m0s
2019/04/28 13:06:03 [INFO] [jdharmon.com] acme: Preparing to solve DNS-01
2019/04/28 13:06:05 [INFO] [jdharmon.com] acme: Trying to solve DNS-01
2019/04/28 13:06:05 [INFO] [jdharmon.com] acme: Checking DNS record propagation using [google-public-dns-a.google.com:53 google-public-dns-b.google.com:53]
2019/04/28 13:06:05 [INFO] Wait for propagation [timeout: 1m0s, interval: 2s]
2019/04/28 13:06:05 [INFO] [jdharmon.com] acme: Waiting for DNS record propagation.
2019/04/28 13:06:15 [INFO] [jdharmon.com] The server validated our request
2019/04/28 13:06:15 [INFO] [jdharmon.com] acme: Cleaning DNS-01 challenge
2019/04/28 13:06:15 [INFO] [*.jdharmon.com, jdharmon.com] acme: Validations succeeded; requesting certificates
2019/04/28 13:06:17 [INFO] [*.jdharmon.com] Server responded with a certificate.
```

**Pull request merged**